pkg/ring/incremental_transfer.go

```diff
@@ -0,0 +1,252 @@
+// ingester address.
+type transferWorkload map[string][]TokenRange
+
+// Do executes functions for the transfer workload in parallel.
+func (wl transferWorkload) Do(f func(addr string, ranges []TokenRange)) {
+	var wg sync.WaitGroup
+	for addr, ranges := range wl {
+		wg.Add(1)
+		go func(addr string, ranges []TokenRange) {
+			defer wg.Done()
+			f(addr, ranges)
+		}(addr, ranges)
+	}
+	wg.Wait()
+}
+
+// findTransferWorkload finds all ranges that should be sent/requested against
+// a target ingester address.
+func (i *Lifecycler) findTransferWorkload(d *Desc, token StatefulToken) (transferWorkload, bool) {
+	ret := make(transferWorkload)
+	rf := i.cfg.RingConfig.ReplicationFactor
+
+	op := Read
+	if token.State == LEAVING {
+		op = Write
+	}
+
+	healthy := d.HealthChecker(op, i.cfg.RingConfig.HeartbeatTimeout)
+
+	var lastErr error
+
+	for replica := 0; replica < rf; replica++ {
+		// Find the ranges which our token is holding data for as one of the replicas.
+		ends, err := d.Predecessors(token, replica, healthy)
+		if err != nil {
+			lastErr = err
+			continue
+		}
+
+		for _, end := range ends {
+			// Find the starting value for the range.
+			start, err := d.Neighbor(end.StatefulToken(), -1, false, healthy)
+			if err != nil {
+				lastErr = err
+				continue
+			}
+
+			// The target token's ingester is the ingester that either has data we
+			// want or should have data we'll send. When joining, it's the old
+			// end of the replica set. When leaving, it's the new end of the replica set.
+			target, err := d.Neighbor(start.StatefulToken(), rf, false, healthy)
+			if err != nil {
+				lastErr = err
+				continue
+			}
+
+			// If our local ingester is already in the replica set, we don't need to
+			// transfer. Our current token will never be considered since it's only
+			// ever LEAVING or JOINING at this point.
+			checkRange := RangeOptions{
+				Range: TokenRange{From: end.Token, To: target.Token},
+
+				ID:          i.ID,
+				IncludeFrom: true,
+				IncludeTo:   true,
+			}
+			if ok, _ := d.InRange(checkRange, healthy); ok {
+				continue
+			}
+
+			addr := d.Ingesters[target.Ingester].Addr
+			ret[addr] = append(ret[addr], TokenRange{From: start.Token, To: end.Token})
+		}
+	}
+
+	if lastErr != nil {
+		level.Error(util.Logger).Log("msg", fmt.Sprintf("failed to find complete transfer set for %d", token.Token), "err", lastErr)
+	}
+	return ret, lastErr == nil
+}
+
+// joinIncrementalTransfer will attempt to incrementally obtain chunks from
+// neighboring lifecyclers that contain data for token ranges they will
+// no longer receive writes for.
+func (i *Lifecycler) joinIncrementalTransfer(ctx context.Context) error {
+	// Make sure that we set all tokens to ACTIVE, even when we fail.
+	defer func() {
+		i.setTokens(i.getTransitioningTokens())
+		i.setTokensState(ACTIVE)
+		i.changeState(ctx, ACTIVE)
+	}()
+
+	r := i.getDenormalisedRing(ctx)
+
+	replicationFactor := i.cfg.RingConfig.ReplicationFactor
```

Comment:
This is an optimization for cases when number of active ingesters is low. Let's put some comment here.